pkg/fqdn/cache.go

```diff
@@ -775,9 +775,34 @@ func (zombies *DNSZombieMappings) GC() (alive, dead []*DNSZombieMapping) {
 	zombies.Lock()
 	defer zombies.Unlock()
 
+	// a name is alive if at least one of the IPs that resolve to it are alive
+	var aliveNames map[string]struct{} = map[string]struct{}{}
+	for _, z := range zombies.deletes {
+		if zombies.isAlive(z) {
+			for _, name := range z.Names {
+				aliveNames[name] = struct{}{}
+			}
+		}
+	}
+
+	// a zombie is not dead if it's alive or its name is alive
+	isZombieNotDead := func(zombie *DNSZombieMapping) bool {
+		if zombies.isAlive(zombie) {
+			return true
+		}
+
+		for _, name := range zombie.Names {
+			if _, ok := aliveNames[name]; ok {
+				return true
+			}
+		}
+
+		return false
+	}
+
 	// Collect zombies we can delete
 	for _, zombie := range zombies.deletes {
-		if zombies.isAlive(zombie) {
+		if isZombieNotDead(zombie) {
```

Comment:
Having two functions, one `isZombieNotDead` and one `zombie.isAlive()` is very confusing. Is there a reason this logic isn't being integrated into `zombie.isAlive()`?

Reply:
I guess the main reason right now is to not iterate from scratch every time `isAlive()` is called. I agree though that the naming is unfortunate.

Reply:
I agree that the name is not great, but my motivation was to separate the two different cases:
 * zombie is alive -- i.e., the connection is alive
 * zombie will not be considered dead -- i.e., either zombie is alive or there is another zombie with the same name that is alive

Hence the comment above:
> 	// a zombie is not dead if it's alive or its name is alive

Overall, the original intention of the code was against what we are doing in this patch, and I wanted to keep the original intention intact (hence the awkward comment on a different file) so that eventually we can refactor it to match the current intent.

Reply:
I believe that the nuance of alive vs. not dead will be completely lost on the next person to read the code.

This seems like it would capture the intent of your response though:

`isAlive()` -> `isConnectionAlive()`
`isZombieNotDead()` -> `isAlive()`

Reply:
Agreed, that's better!

Pushed a new version with this change, which also uncovered that we only printed the alive zombies based on the new definition. Now we print the alive zombies with the new definition.

Reply:
@joestringer Names look good from your side? We landed on isConnectionAlive and isZombieAlive it looks like.